### PR TITLE
[#15] Enter passwords to load levels!

### PR DESCRIPTION
**Pull request by cdroulers**
Missing a test for goto level with password.

### CODE REVIEW

**Comment by cdroulers** [comment]
Bugs found:

* using shift+left after loading a further level will show the previous level!
* the top display doesn't update to the latest level

This is all due to passing around canvas objects and not have a proper pattern for UI vs. controllers.
Look into implementing a simple MVC or MVVM pattern with pub/sub or something.

**Comment by olafleur** [comment]
Really great job !

Suggestion : maybe clear the text box when submitting ?
When you reopen the dialog for entering a password, the old one is still there.

File: src/block-fly/app.ts [hunk 1]
```diff
@@ -1,58 +1,53 @@
...
 require("material-design-lite/material.min"); // tslint:disable-line no-require-imports no-var-requires
 
-const canvas = document.getElementById("root") as HTMLCanvasElement;
-const canvasTitle = document.querySelector("#level-indicator > div:first-child") as HTMLDivElement;
-
 setupDialogs();
 
-const parser = new BoardParser();
+const controller = new Controller();
+console.log("Started controller", controller);
```
Comment (olafleur) [inline src/block-fly/app.ts]: is this intended ?

File: src/block-fly/infrastructure/controller.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import * as Events from "./events";
+import { showMessage, showErrorMessage } from "../display/messageDisplay";
+import { writeToCanvas, getActualDimensions } from "../display/canvasDisplay";
+import { IViewport, getViewport } from "../display/viewport";
+
+const parser = new BoardParser();
+
+export default class Controller {
+  private _canvas: HTMLCanvasElement;
+  private canvasTitle: HTMLDivElement;
```
Comment (olafleur) [inline src/block-fly/infrastructure/controller.ts:13]: shouldn't you be consistent with the underscore before private variables?

**Review by olafleur — APPROVED** [review]

File: src/block-fly/app.ts (hunk 1 above)
Comment (cdroulers) [reply src/block-fly/app.ts]: I'll remove it!

File: src/block-fly/infrastructure/controller.ts (hunk 2 above)
Comment (cdroulers) [reply src/block-fly/infrastructure/controller.ts:13]: Absolutely!

**Pull request merged**